xbmc/addons/Addon.cpp

```diff
@@ -32,57 +32,93 @@
 
 using XFILE::CDirectory;
 using XFILE::CFile;
+using namespace KODI::ADDONS;
 
 namespace ADDON
 {
 
 CAddon::CAddon(const AddonInfoPtr& addonInfo, TYPE addonType)
   : m_addonInfo(addonInfo),
-    m_userSettingsPath(),
-    m_loadSettingsFailed(false),
-    m_hasUserSettings(false),
     m_profilePath(StringUtils::Format("special://profile/addon_data/{}/", m_addonInfo->ID())),
-    m_settings(nullptr),
     m_type(addonType == ADDON_UNKNOWN ? addonInfo->MainType() : addonType)
 {
-  m_userSettingsPath = URIUtils::AddFileToFolder(m_profilePath, "settings.xml");
 }
 
 /**
  * Settings Handling
  */
+
+std::vector<uint32_t> CAddon::GetKnownInstanceIds() const
+{
+  return m_addonInfo->GetKnownInstanceIds();
+}
+
+bool CAddon::SupportInstanceSettings() const
+{
+  return m_addonInfo->SupportInstanceSettings();
+}
+
+bool CAddon::DeleteInstanceSettings(uint32_t instance)
+{
+  auto itr = m_settings.find(instance);
+  if (instance == KODI::ADDONS::ADDON_SETTINGS_ID || itr == m_settings.end())
+    return false;
+
+  if (CFile::Exists(itr->second.m_userSettingsPath))
+    CFile::Delete(itr->second.m_userSettingsPath);
```

Comment:
The old problem: checking a non-lockable resource for existence guarantees nothing. You must implement proper error handling for the `Delete`call, because the fiel might get deleted right after checking existense and before calling `Delete`. So, omit the `Exist` call and react properly on `not-exist`error on `Deelte`call.